lucene/core/src/java/org/apache/lucene/index/IndexWriter.java

```diff
@@ -3265,24 +3269,40 @@ public void mergeFinished(boolean committed, boolean segmentDropped) throws IOEx
                           }
                           // Construct a OneMerge that applies to toCommit
                           MergePolicy.OneMerge applicableMerge = new MergePolicy.OneMerge(toCommitMergedAwaySegments);
-                          applicableMerge.info = info.clone();
-                          long segmentCounter = Long.parseLong(info.info.name.substring(1), Character.MAX_RADIX);
+                          applicableMerge.info = origInfo;
+                          long segmentCounter = Long.parseLong(origInfo.info.name.substring(1), Character.MAX_RADIX);
                           committingSegmentInfos.counter = Math.max(committingSegmentInfos.counter, segmentCounter + 1);
                           committingSegmentInfos.applyMergeChanges(applicableMerge, false);
                         }
                         toWrap.mergeFinished(committed, false);
                         super.mergeFinished(committed, segmentDropped);
                       }
 
+                      @Override
+                      void onMergeCommit() {
+                        origInfo = this.info.clone();
```

Comment:
Whoa!  This is the functional part of the fix?  Because we snapshot the `SegmentCommitInfo` *before* carrying over deletes that happened while the merge was running?  Very cool.

But, how do we ensure that the prior live docs (without concurrent deletions carried over) are written for this merged segment, and matching `origInfo`'s `delGen`?  Normally `commitMergedDeletesAndUpdates` would do that, except it is carrying over the concurrent deletions, which we need to avoid.  Oh, wait, sorry: such deletions are of course merged away!  So, we want 100% live docs in the newly merged segment.  Tricky.

We probably also need a test case confirm that (concurrent) DV updates are also transactional?  I think this same approach should work for DV updates too?  We will keep only the DV updates that got merged away, and any new ones that we carry over will not be visible in the merged segment because you are cloning the `SegmentCommitInfo` before `commitMerge` runs?